sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala

```diff
@@ -2007,4 +2007,76 @@ class RegexUnsupportedException(message: String, index: Option[Int])
       case _ => message
     }
   }
-}
\ No newline at end of file
+}
+
+sealed trait RegexOptimizationType
+object RegexOptimizationType {
+  case class StartsWith(pattern: String) extends RegexOptimizationType
+  case class Contains(pattern: String) extends RegexOptimizationType
+  case object NoOptimization extends RegexOptimizationType
+}
+
+object RegexRewriteUtils {
+  private def isSimplePattern(astLs: Seq[RegexAST]): Boolean = {
+    astLs.forall {
+      case RegexChar('^') | RegexChar('$') | RegexChar('.') => false
+      case RegexChar(_) => true
+      case _ => false
+    }
+  }
+
+  private def isWildcard(ast: RegexAST): Boolean = {
+    ast match {
+      case RegexRepetition(RegexChar('.'), SimpleQuantifier('*')) => true
+      case RegexSequence(parts) if parts.forall(isWildcard) => true
+      case RegexGroup(_, term, _) if isWildcard(term) => true
+      case _ => false
+    }
+  }
+
+  private def stripLeadingWildcards(astLs: Seq[RegexAST]): Seq[RegexAST] = astLs match {
+    case (RegexChar('^') | RegexEscaped('A')) :: tail  =>
+      tail.dropWhile(isWildcard)
+    case _ => astLs.dropWhile(isWildcard)
+  }
+
+  private def stripTailingWildcards(astLs: Seq[RegexAST]): Seq[RegexAST] = {
+    astLs.reverse.dropWhile(isWildcard).reverse
+  }
+
+  private def RegexCharsToString(chars: Seq[RegexAST]): String = {
+    chars.map {
+      case RegexChar(ch) => ch
+      case _ => throw new IllegalArgumentException("Invalid character")
+    }.mkString
+  }
+
+  def matchSimplePattern(ast: RegexAST): RegexOptimizationType = {
+    ast.children().toList match {
+      case (RegexChar('^') | RegexEscaped('A')) :: RegexGroup(_, RegexSequence(parts), None) :: rest
+          if isSimplePattern(parts.toList) && rest.forall(isWildcard) => {
+        // ^(pattern).* => startsWith pattern
+        RegexOptimizationType.StartsWith(RegexCharsToString(parts.toList))
+      }
+      case (RegexChar('^') | RegexEscaped('A')) :: ast 
+          if isSimplePattern(stripTailingWildcards(ast)) => {
+        // ^pattern.* => startsWith pattern
+        RegexOptimizationType.StartsWith(RegexCharsToString(stripTailingWildcards(ast)))
+      }
+      case noStartsWithAst => stripLeadingWildcards(noStartsWithAst) match {
+        case RegexGroup(_, RegexSequence(parts), None) :: rest
+            if isSimplePattern(parts.toList) && rest.forall(isWildcard) => {
+          // (pattern).* => contains pattern
+          RegexOptimizationType.Contains(RegexCharsToString(parts.toList))
+        }
+        case ast if isSimplePattern(stripTailingWildcards(ast)) => {
+          // pattern.* => contains pattern
+          RegexOptimizationType.Contains(RegexCharsToString(stripTailingWildcards(ast)))
+        }
+        case _ => {
+          RegexOptimizationType.NoOptimization
+        }
+      }
+    }
+  }
```

Comment:
after you made the API Seq-based you should be able to get rid of `toList` conversions.

Reply:
done.

Reply:
From scala 2.13: 
```
scala.Seq[+A] is now an alias for scala.collection.immutable.Seq[A] (instead of scala.collection.Seq[A]). Note that this also changes the type of Scala varargs methods.
```
So there are some errors in 2.13 build
```
Error: ] /home/runner/work/spark-rapids/spark-rapids/sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:2057: type mismatch;
 found   : scala.collection.mutable.ListBuffer[com.nvidia.spark.rapids.RegexAST]
 required: Seq[com.nvidia.spark.rapids.RegexAST]
```

Now updated those APIs to `collection.Seq` per [migration doc](https://docs.scala-lang.org/overviews/core/collections-migration-213.html).